pkg/identity/principal.go

```diff
@@ -0,0 +1,29 @@
+// Copyright 2022 The Sigstore Authors.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package identity
+
+import (
+	"context"
+	"crypto/x509"
+)
+
+type Principal interface {
+	// URI, email etc of principal (usually matches `sub` from ID token)
+	Name(ctx context.Context) string
+
+	// Embed all SubjectNameAlt and custom x509 extension information into
+	// certificate.
+	Embed(ctx context.Context, cert x509.Certificate) (x509.Certificate, error)
```

Comment:
Is there an advantage to a pointer? I prefer a pure function over mutation where possible

Reply:
I prefer it because it signals that the initial certificate will get mutated. Otherwise, the test needs to demonstrate that nothing else was mutated but the fields passed in.
Also for performance - Certs can be large, so avoiding copying is preferred.

Unfortunately there's no central style guide for Go, so either is acceptable, but fwiw, when working with certs, I almost always see them passed by pointer.

Reply:
Alright I'll switch the signature to a mutating one